File: notifications/Actors/SubscriptionActor.cs
extern alias Shaded;
using System.Collections.Generic;
using System.Threading.Tasks;
using Dapr.Actors;
using Dapr.Actors.Runtime;
using Shaded.Dapr.Client;
using Shared;

namespace Notifications.Actors
{
    /// <summary>
    /// An actor that is responsible for handling the subscriptions in the state store.
    /// </summary>
    public interface ISubscriptionActor : IActor
    {
        /// <summary>
        /// Adds the given subscriptions to the state store.
        /// </summary>
        /// <param name="subscription">The subscription data.</param>
        /// <returns>A task that represents the asynchronous operation.</returns>
        Task Subscribe(Subscription subscription);

        /// <summary>
        /// Removes the subscription for the given email from the state store.
        /// </summary>
        /// <param name="email">The email to unsubscribe.</param>
        /// <returns>A task that represents the asynchronous operation.</returns>
        Task Unsubscribe(string email);
    }

    /// <summary>
    /// Concrete subscription actor implementation.
    ///
    /// Storing all subscriptions under a single key, may be an overhead for a large set of subscriptions. For
    /// production usage, an other storage format may be better.
    /// </summary>
    [Actor(TypeName = SubscriptionActor.Name)]
    public class SubscriptionActor : Actor, ISubscriptionActor
    {
        /// <summary>
        /// The name of the actor.
        /// </summary>
        public const string Name = "SubscriptionActor";

        /// <summary>
        /// The ID of the actor.
        /// </summary>
        public static ActorId ID = new ActorId("subscription");

        private readonly DaprClient _daprClient;

        /// <summary>
        /// Initializes a new instance of the <see cref="SubscriptionActor"/> class.
        /// </summary>
        /// <param name="actorService">The <see cref="P:Dapr.Actors.Runtime.Actor.ActorService" /> that will host this actor instance.</param>
        /// <param name="actorId">ID for the actor.</param>
        /// <param name="daprClient">A dapr client instance.</param>
        public SubscriptionActor(ActorService actorService, ActorId actorId, DaprClient daprClient)
            : base(actorService, actorId)
        {
            _daprClient = daprClient;
        }

        /// <inheritdoc/>
        async public Task Subscribe(Subscription subscription)
        {
            var subscriptions = await _daprClient.GetStateAsync<Dictionary<string, Subscription>>(
                Config.StateStore.name,
                Config.StateStore.subscriptions);

            Subscription? value = subscriptions.GetValueOrDefault(subscription.Email);
            // Add a new subscription if not exists or updates the subscription if the name has changed
            if (!value.HasValue || value.Value.Name != subscription.Name)
            {
                subscriptions[subscription.Email] = subscription;

                await _daprClient.SaveStateAsync(
                    Config.StateStore.name,
                    Config.StateStore.subscriptions,
                    subscriptions);
            }
        }

        /// <inheritdoc/>
        async public Task Unsubscribe(string email)
        {
            var subscriptions = await _daprClient.GetStateAsync<Dictionary<string, Subscription>>(
                Config.StateStore.name,
                Config.StateStore.subscriptions);

            var updated = subscriptions.Remove(email);
            if (updated)
            {
                await _daprClient.SaveStateAsync(
                    Config.StateStore.name,
                    Config.StateStore.subscriptions,
                    subscriptions);
            }

        }
    }
}

File: subscriptions/Actors/SubscriptionActor.fs
namespace Subscriptions.Actors

open Dapr.Client
open Dapr.Actors
open Dapr.Actors.Runtime
open FSharpx.Control
open System.Threading.Tasks
open Shared.Config
open Shared.Extensions
open Shared

module SubscriptionActor =

    /// <summary>
    /// The name of the actor.
    /// </summary>
    [<Literal>]
    let Name = "SubscriptionActor"

    /// <summary>
    /// The ID of the actor.
    /// </summary>
    let ID = ActorId("subscription")

/// <summary>
/// An actor that is responsible for handling the subscriptions in the state store.
/// </summary>
type ISubscriptionActor =
    inherit IActor

    /// <summary>
    /// Adds the given subscriptions to the state store.
    /// </summary>
    /// <param name="subscription">The subscription data.</param>
    /// <returns>A task that represents the asynchronous operation.</returns>
    abstract Subscribe: subscription: Subscription -> Task<Subscriptions>

    /// <summary>
    /// Removes the subscription for the given email from the state store.
    /// </summary>
    /// <param name="email">The email to unsubscribe.</param>
    /// <returns>A task that represents the asynchronous operation.</returns>
    abstract Unsubscribe: email: string -> Task<Subscriptions>

/// <summary>
/// Concrete subscription actor implementation.
///
/// Storing all subscriptions under a single key, may be an overhead for a large set of subscriptions. For
/// production usage, an other storage format may be better.
/// </summary>
[<Actor(TypeName = SubscriptionActor.Name)>]
type SubscriptionActor(actorService: ActorService, actorId: ActorId, daprClient: DaprClient) =
    inherit Actor(actorService, actorId)

    interface ISubscriptionActor with

        /// <inheritdoc/>
        member _.Subscribe(subscription: Subscription) =
            async {
                let! subscriptions =
                    daprClient.GetStateAsyncOr(StateStore.name, Apps.subscriptions, Map.empty<string, Subscription>)

                let maybeValue = subscriptions.TryFind(subscription.Email)

                return!
                    match maybeValue with
                    | Some(value) when value.Name = subscription.Name ->
                        async { return subscriptions }
                    | _ ->
                        let updatedSubscriptions = subscriptions.Add(subscription.Email, subscription)

                        daprClient.SaveStateAsync(StateStore.name, Apps.subscriptions, updatedSubscriptions)
                        |> Async.AwaitTask
                        |> Async.map (fun _ -> updatedSubscriptions)
            }
            |> Async.StartAsTask

        /// <inheritdoc/>
        member _.Unsubscribe(email: string) =
            async {
                let! subscriptions =
                    daprClient.GetStateAsyncOr(StateStore.name, Apps.subscriptions, Map.empty<string, Subscription>)

                return!
                    match subscriptions.ContainsKey(email) with
                    | true ->
                        daprClient.SaveStateAsync(StateStore.name, Apps.subscriptions, subscriptions.Remove(email))
                        |> Async.AwaitTask
                        |> Async.map (fun _ -> subscriptions)
                    | false -> async { return subscriptions }
            }
            |> Async.StartAsTask
